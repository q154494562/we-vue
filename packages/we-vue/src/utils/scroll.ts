type ElementOrWindow = HTMLElement | Window

/**
 * 找到最近的触发滚动事件的元素
 * @param {Element} element
 * @param {Element} rootParent
 * @returns {Element | window}
 */
export function getScrollEventTarget (element: HTMLElement, rootParent: ElementOrWindow = window) {
  let node = element
  while (
    node &&
    node.tagName !== 'HTML' &&
    node.tagName !== 'BODY' &&
    node.nodeType === 1 &&
    node !== rootParent
  ) {
    const { overflowY } = window.getComputedStyle(node)
    if (overflowY === 'scroll' || overflowY === 'auto') {
      return node
    }
    node = <HTMLElement>node.parentNode
  }
  return rootParent
}

// 获取滚动高度
export function getScrollTop (element: ElementOrWindow): number {
  return 'scrollTop' in element ? element.scrollTop : element.pageYOffset
}

// 设置滚动高度
export function setScrollTop (element: ElementOrWindow, value: number) {
  'scrollTop' in element ? (element.scrollTop = value) : element.scrollTo(element.scrollX, value)
}

// 获取元素距离顶部高度
export function getElementTop (element: ElementOrWindow) {
  return (
    (element === window ? 0 : (element as HTMLElement).getBoundingClientRect().top) + getScrollTop(window)
  )
}

export function getVisibleHeight (element: ElementOrWindow) {
  return element === window
    ? element.innerHeight
    : (element as HTMLElement).getBoundingClientRect().height
}
