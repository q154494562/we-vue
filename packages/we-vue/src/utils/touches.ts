function getTouch (event: TouchEvent): Touch {
  return event.changedTouches[0] || event.touches[0]
}

export { getTouch }
