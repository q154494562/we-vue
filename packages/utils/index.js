import Vue from 'vue'
import { getTouch } from './touches'

const isServer = Vue.prototype.$isServer

function isDef (value) {
  return value !== undefined && value !== null
}

function isObj (x) {
  const type = typeof x
  return x !== null && (type === 'object' || type === 'function')
}

export {
  isServer,
  getTouch,
  isDef,
  isObj
}
