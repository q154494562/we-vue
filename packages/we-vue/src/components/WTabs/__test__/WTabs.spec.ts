import { mount, shallowMount } from '@vue/test-utils'
import Tabs from '../'
import Tab from '../../WTab'
import { horizontalDrag } from '@/test/unit/utils'

describe('tabs', () => {
  test('create', () => {
    const wrapper = shallowMount(Tabs, {
      propsData: {},
    })

    expect(wrapper.name()).toBe('w-tabs')
    expect(wrapper.classes()).toContain('wv-tabs')
  })

  test('create with tab items', () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      propsData: {},
      slots: {
        default: [Tab, Tab, Tab],
      },
    })

    expect(wrapper.html()).toMatchSnapshot()

    expect(wrapper.name()).toBe('w-tabs')
    expect(wrapper.classes()).toContain('wv-tabs')
  })

  test('lineHeight property', async () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      propsData: {
        lineHeight: 10,
      },
      slots: {
        default: [Tab, Tab, Tab],
      },
    })

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()

    expect((wrapper.vm.lineStyle as any).height).toBe('10px')
    expect((wrapper.vm.lineStyle as any).borderRadius).toBe('10px')
  })

  test('color property', async () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      propsData: {
        color: 'red',
      },
      slots: {
        default: [Tab, Tab, Tab],
      },
    })

    await wrapper.vm.$nextTick()

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('getTabStyle method', async () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      propsData: {
        color: 'red',
        type: 'card',
        titleActiveColor: '#fff',
        titleInactiveColor: '#000',
      },
      slots: {
        default: [Tab, Tab, Tab],
      },
    })

    await wrapper.vm.$nextTick()

    const tab1 = wrapper.findAll(Tab).at(0)

    expect(wrapper.vm.getTabStyle(tab1.vm as any, 0)).toEqual({
      backgroundColor: 'red',
      borderColor: 'red',
      color: '#fff',
    })
  })

  test('click tab items', async () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      propsData: {},
      slots: {
        default: [Tab, Tab, Tab],
      },
    })

    await wrapper.vm.$nextTick()

    wrapper
      .findAll('.wv-tab')
      .at(1)
      .trigger('click')
  })

  test('swipe content', async () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      propsData: {
        swipeable: true,
      },
      slots: {
        default: [Tab, Tab, Tab, Tab, Tab],
      },
    })

    const spy = jest.spyOn(wrapper.vm, 'setCurActive')
    await wrapper.vm.$nextTick()

    // drag to left
    horizontalDrag(wrapper.vm.$refs.content, 0, -1000)
    expect(spy).toHaveBeenCalledWith(1)

    await wrapper.vm.$nextTick()

    // drag to right
    horizontalDrag(wrapper.vm.$refs.content, 0, 1000)
    expect(spy).toHaveBeenLastCalledWith(0)
  })

  test('watch active', async () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      slots: {
        default: [Tab, Tab, Tab],
      },
    })

    const spy = jest.spyOn(wrapper.vm, 'setLine')
    await wrapper.vm.$nextTick()

    wrapper.setProps({
      active: 1,
    })

    expect(spy).toHaveBeenCalled()
  })

  test('watch color', async () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      slots: {
        default: [Tab, Tab, Tab],
      },
    })

    const spy = jest.spyOn(wrapper.vm, 'setLine')
    await wrapper.vm.$nextTick()
    wrapper.setProps({
      color: '#f00',
    })

    expect(spy).toHaveBeenCalled()
  })

  test('watch sticky', async () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      slots: {
        default: [Tab, Tab, Tab],
      },
    })

    const spy = jest.spyOn(wrapper.vm, 'handlers')
    await wrapper.vm.$nextTick()
    wrapper.setProps({
      sticky: true,
    })

    expect(spy).toHaveBeenCalled()
  })

  test('watch swipeable', async () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      slots: {
        default: [Tab, Tab, Tab],
      },
    })

    const spy = jest.spyOn(wrapper.vm, 'handlers')
    await wrapper.vm.$nextTick()
    wrapper.setProps({
      swipeable: true,
    })

    expect(spy).toHaveBeenCalled()
  })

  test('beforeDestroy hooks', () => {
    const wrapper = mount(Tabs, {
      attachToDocument: true,
      slots: {
        default: [Tab, Tab, Tab],
      },
    })

    const spy = jest.spyOn(wrapper.vm, 'handlers')
    wrapper.vm.$destroy()
    expect(spy).toHaveBeenCalled()
  })
})
