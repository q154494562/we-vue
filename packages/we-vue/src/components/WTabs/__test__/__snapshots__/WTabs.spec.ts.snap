// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tabs color property 1`] = `
<div class="wv-tabs wv-tabs--line">
  <div class="wv-tabs__wrap wv-hairline--top-bottom">
    <div class="wv-tabs__nav wv-tabs__nav--line" style="border-color: red;">
      <div class="wv-tabs__line" style="background-color: red; width: 0px; transform: translateX(0px); transition-duration: 0.2s;"></div>
      <div class="wv-tab wv-tab--active"><span class="wv-ellipsis"></span></div>
      <div class="wv-tab"><span class="wv-ellipsis"></span></div>
      <div class="wv-tab"><span class="wv-ellipsis"></span></div>
    </div>
  </div>
  <div class="wv-tabs__content">
    <div class="wv-tab__pane"></div>
    <div class="wv-tab__pane" style="display: none;"></div>
    <div class="wv-tab__pane" style="display: none;"></div>
  </div>
</div>
`;

exports[`tabs create with tab items 1`] = `
<div class="wv-tabs wv-tabs--line">
  <div class="wv-tabs__wrap wv-hairline--top-bottom">
    <div class="wv-tabs__nav wv-tabs__nav--line">
      <div class="wv-tabs__line"></div>
      <div class="wv-tab wv-tab--active"><span class="wv-ellipsis"></span></div>
      <div class="wv-tab"><span class="wv-ellipsis"></span></div>
      <div class="wv-tab"><span class="wv-ellipsis"></span></div>
    </div>
  </div>
  <div class="wv-tabs__content">
    <div class="wv-tab__pane"></div>
    <div class="wv-tab__pane" style="display: none;"></div>
    <div class="wv-tab__pane" style="display: none;"></div>
  </div>
</div>
`;

exports[`tabs lineHeight property 1`] = `
<div class="wv-tabs wv-tabs--line">
  <div class="wv-tabs__wrap wv-hairline--top-bottom">
    <div class="wv-tabs__nav wv-tabs__nav--line">
      <div class="wv-tabs__line" style="width: 0px; transform: translateX(0px); transition-duration: 0.2s; height: 10px; border-radius: 10px;"></div>
      <div class="wv-tab wv-tab--active"><span class="wv-ellipsis"></span></div>
      <div class="wv-tab"><span class="wv-ellipsis"></span></div>
      <div class="wv-tab"><span class="wv-ellipsis"></span></div>
    </div>
  </div>
  <div class="wv-tabs__content">
    <div class="wv-tab__pane"></div>
    <div class="wv-tab__pane" style="display: none;"></div>
    <div class="wv-tab__pane" style="display: none;"></div>
  </div>
</div>
`;
